### PR TITLE
E2E test for vsphere shared volume driver

### PR DESCRIPTION
**Pull request by lipingxue**
**This PR includes:**
1. E2E test to test volume create with various valid/invalid option
2. E2E test for basic life cycle, which include create/get/inspect/rm now.  Mount/unmount test will be added later after mount/unmount is available.
3. necessary changes in Makefile

**Test:**
prerequisites: need to install vsphere shared volume driver and set alias to "vshared" on all docker hosts.

```
lipingx-m01:docker-volume-vsphere lipingx$ make test-e2e-shared 
/Library/Developer/CommandLineTools/usr/bin/make --directory=client_plugin test-e2e-shared

=> Running target test-e2e-shared Tue Aug 8 10:50:18 PDT 2017

../misc/scripts/build.sh test-e2e-runonce-shared
make: Entering directory `/go/src/github.com/vmware/docker-volume-vsphere/client_plugin'

=> Running target test-e2e-runonce-shared Tue Aug 8 17:50:20 UTC 2017

go test -v -timeout 40m -tags "runonceshared" github.com/vmware/docker-volume-vsphere/tests/e2e
=== RUN   Test
2017/08/08 17:50:23 START: BasicSharedTestSuite.TestVolumeLifecycle
2017/08/08 17:50:23 Creating shared volume [BasicSharedTestSuite.TestVolumeLifecycle_volume_902000] on VM [10.160.100.14]
2017/08/08 17:50:29 Checking volume [BasicSharedTestSuite.TestVolumeLifecycle_volume_902000] availability from VM [10.160.100.14]
2017/08/08 17:50:30 GetSharedVolumeStatusHost: cmd[]%!(EXTRA string=docker volume inspect  --format "{{index .Status \"Volume Status\"}}" BasicSharedTestSuite.TestVolumeLifecycle_volume_902000)
2017/08/08 17:50:30 GetSharedVolumeStatusHost return out[%s] for volume %s Ready BasicSharedTestSuite.TestVolumeLifecycle_volume_902000
2017/08/08 17:50:30 Checking volume [BasicSharedTestSuite.TestVolumeLifecycle_volume_902000] availability from VM [10.160.100.14]
2017/08/08 17:50:31 Destroying volume [BasicSharedTestSuite.TestVolumeLifecycle_volume_902000]
2017/08/08 17:50:33 Creating shared volume [BasicSharedTestSuite.TestVolumeLifecycle_volume_902000] on VM [10.160.99.2]
2017/08/08 17:50:39 Checking volume [BasicSharedTestSuite.TestVolumeLifecycle_volume_902000] availability from VM [10.160.99.2]
2017/08/08 17:50:40 GetSharedVolumeStatusHost: cmd[]%!(EXTRA string=docker volume inspect  --format "{{index .Status \"Volume Status\"}}" BasicSharedTestSuite.TestVolumeLifecycle_volume_902000)
2017/08/08 17:50:40 GetSharedVolumeStatusHost return out[%s] for volume %s Ready BasicSharedTestSuite.TestVolumeLifecycle_volume_902000
2017/08/08 17:50:40 Checking volume [BasicSharedTestSuite.TestVolumeLifecycle_volume_902000] availability from VM [10.160.99.2]
2017/08/08 17:50:41 Destroying volume [BasicSharedTestSuite.TestVolumeLifecycle_volume_902000]
2017/08/08 17:50:43 END: BasicSharedTestSuite.TestVolumeLifecycle
2017/08/08 17:50:43 START: VolumeCreateSharedTestSuite.TestInvalidName
2017/08/08 17:50:43 Creating shared volume [volume/abc] on VM [10.160.100.14]
2017/08/08 17:50:43 Creating shared volume [Volume-000000] on VM [10.160.100.14]
2017/08/08 17:50:43 Creating shared volume [BpLnfgDsc2WD8F2qNfHK5a84jjJkwzDkh9h2fhfUVuS9jZ8uVbhV3vC5AWX39IVUWSP2NcHciWvqZTa2N95RxRTZHWUsaD6HEdz0T] on VM [10.160.100.14]
2017/08/08 17:50:43 Creating shared volume [Volume_volume_305432@invalidDatastore] on VM [10.160.100.14]
2017/08/08 17:50:45 END: VolumeCreateSharedTestSuite.TestInvalidName
2017/08/08 17:50:45 START: VolumeCreateSharedTestSuite.TestInvalidOptions
2017/08/08 17:50:45 Creating shared volume [option_volume_974534] with options [ -o access=read-write-both] on VM [10.160.100.14]
2017/08/08 17:50:45 Creating shared volume [option_volume_752538] with options [ -o diskformat=zeroedthick,thin] on VM [10.160.100.14]
2017/08/08 17:50:45 Creating shared volume [option_volume_931047] with options [ -o size=100mbb] on VM [10.160.100.14]
2017/08/08 17:50:45 Creating shared volume [option_volume_200376] with options [ -o size=100gbEE] on VM [10.160.100.14]
2017/08/08 17:50:45 Creating shared volume [option_volume_892013] with options [ -o sizes=100mb] on VM [10.160.100.14]
2017/08/08 17:50:45 Creating shared volume [option_volume_647695] with options [ -o diskformat=zeroedthickk] on VM [10.160.100.14]
2017/08/08 17:50:45 Creating shared volume [option_volume_728097] with options [ -o fstype=xfs_ext] on VM [10.160.100.14]
2017/08/08 17:50:45 Creating shared volume [option_volume_466696] with options [ -o access=read-write-both] on VM [10.160.100.14]
2017/08/08 17:50:45 Creating shared volume [option_volume_1005756] with options [ -o access=write-only] on VM [10.160.100.14]
2017/08/08 17:50:49 END: VolumeCreateSharedTestSuite.TestInvalidOptions
2017/08/08 17:50:49 START: VolumeCreateSharedTestSuite.TestValidName
2017/08/08 17:50:49 Creating shared volume ["Volume Space"] on VM [10.160.100.14]
2017/08/08 17:50:49 Creating shared volume [mSzO1fCUWooVQlNpimWSFrLHSXg6xXya0iNuDQLZgAJ5UdCvLSJXbf1NE2wVqDumwKRWMjF7qCNqL5LNJ2ZPUpYZt] on VM [10.160.100.14]
2017/08/08 17:50:49 Creating shared volume [abc_volume_871177@sharedVmfs-0] on VM [10.160.100.14]
2017/08/08 17:50:49 Creating shared volume [Volume-0000000] on VM [10.160.100.14]
2017/08/08 17:50:49 Creating shared volume [abc_volume_1084464@@@@sharedVmfs-0] on VM [10.160.100.14]
2017/08/08 17:50:49 Creating shared volume [Volume-你_volume_831293] on VM [10.160.100.14]
2017/08/08 17:50:49 Creating shared volume [Volume-0000000-****-###] on VM [10.160.100.14]
2017/08/08 17:50:49 Creating shared volume [Volume-00000] on VM [10.160.100.14]
2017/08/08 17:51:07 Checking volume [Volume-0000000 abc_volume_1084464@@@@sharedVmfs-0 mSzO1fCUWooVQlNpimWSFrLHSXg6xXya0iNuDQLZgAJ5UdCvLSJXbf1NE2wVqDumwKRWMjF7qCNqL5LNJ2ZPUpYZt Volume-0000000-****-### Volume-00000 abc_volume_871177@sharedVmfs-0 "Volume Space" Volume-你_volume_831293] availability from VM [10.160.100.14]
2017/08/08 17:51:07 END: VolumeCreateSharedTestSuite.TestValidName
2017/08/08 17:51:07 Destroying volume [Volume-0000000 abc_volume_1084464@@@@sharedVmfs-0 mSzO1fCUWooVQlNpimWSFrLHSXg6xXya0iNuDQLZgAJ5UdCvLSJXbf1NE2wVqDumwKRWMjF7qCNqL5LNJ2ZPUpYZt Volume-0000000-****-### Volume-00000 abc_volume_871177@sharedVmfs-0 "Volume Space" Volume-你_volume_831293]
2017/08/08 17:51:21 START: VolumeCreateSharedTestSuite.TestValidOptions
2017/08/08 17:51:21 Creating shared volume [option_volume_110623] with options [ -o access=read-write] on VM [10.160.100.14]
2017/08/08 17:51:21 Creating shared volume [option_volume_425789] with options [ -o diskformat=eagerzeroedthick] on VM [10.160.100.14]
2017/08/08 17:51:21 Creating shared volume [option_volume_890580] with options [ -o access=read-only] on VM [10.160.100.14]
2017/08/08 17:51:21 Creating shared volume [option_volume_313277] with options [ -o diskformat=zeroedthick] on VM [10.160.100.14]
2017/08/08 17:51:21 Creating shared volume [option_volume_1086080] with options [ -o diskformat=thin] on VM [10.160.100.14]
2017/08/08 17:51:21 Creating shared volume [option_volume_1005060] with options [ -o attach-as=independent_persistent] on VM [10.160.100.14]
2017/08/08 17:51:21 Creating shared volume [option_volume_748273] with options [ -o fstype=ext4] on VM [10.160.100.14]
2017/08/08 17:51:21 Creating shared volume [option_volume_296877] with options [ -o size=10gb] on VM [10.160.100.14]
2017/08/08 17:51:21 Creating shared volume [option_volume_354728] with options [ -o attach-as=persistent] on VM [10.160.100.14]
2017/08/08 17:51:42 Checking volume [option_volume_296877 option_volume_1086080 option_volume_110623 option_volume_890580 option_volume_748273 option_volume_354728 option_volume_313277 option_volume_425789 option_volume_1005060] availability from VM [10.160.100.14]
2017/08/08 17:51:42 END: VolumeCreateSharedTestSuite.TestValidOptions
2017/08/08 17:51:42 Destroying volume [option_volume_296877 option_volume_1086080 option_volume_110623 option_volume_890580 option_volume_748273 option_volume_354728 option_volume_313277 option_volume_425789 option_volume_1005060]
--- PASS: Test (97.26s)
PASS
OK: 5 passed
ok  	github.com/vmware/docker-volume-vsphere/tests/e2e	97.267s
make: Leaving directory `/go/src/github.com/vmware/docker-volume-vsphere/client_plugin'

```

### CODE REVIEW

File: tests/e2e/volumecreate_shared_test.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+	. "gopkg.in/check.v1"
+)
+
+type VolumeCreateSharedTestSuite struct {
+	config     *inputparams.TestConfig
+	volumeList []string
+}
+
+var (
```
Comment (shivanshu21) [inline tests/e2e/volumecreate_shared_test.go]: Any reason why invalid names are a list while valid name list is returned by a function?

File: tests/e2e/basic_shared_test.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+func (s *BasicSharedTestSuite) SetUpTest(c *C) {
+	s.volName1 = inputparams.GetUniqueVolumeName(c.TestName())
+	s.volName2 = inputparams.GetUniqueVolumeName(c.TestName())
+	s.containerName = inputparams.GetUniqueContainerName(c.TestName())
+}
+
+var _ = Suite(&BasicSharedTestSuite{})
+
+// Test volume lifecycle management on different datastores:
+// VM1 - created on local VMFS datastore
```
Comment (shivanshu21) [inline tests/e2e/basic_shared_test.go]: What is local VMFS?

File: tests/e2e/volumecreate_shared_test.go (hunk 1 above)
Comment (lipingxue) [inline tests/e2e/volumecreate_shared_test.go]: No specific reason. I take the code from original  volumecreate_test.go.

File: tests/e2e/basic_shared_test.go (hunk 2 above)
Comment (lipingxue) [inline tests/e2e/basic_shared_test.go]: The comments is wrong, I check the actual configuration, all VMs are created in the shared datastore. I will remove the comment.

**Comment by lipingxue** [comment]
@shivanshu21 I have addressed/replied your comment, and please take a look.

File: tests/e2e/volumecreate_shared_test.go [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+	misc.LogTestEnd(c.TestName())
+}
+
+// Valid volume creation options
+// 1. size 10gb
+// 2. disk format (thin, zeroedthick, eagerzeroedthick)
+// 3. attach-as (persistent, independent_persistent)
+// 4. fstype ext4 for linux
+// 5. access (read-write, read-only)
```
Comment (luomiao) [inline tests/e2e/volumecreate_shared_test.go]: We should only support size for now.
The other options are not valid for shared plugin.

Comment (lipingxue) [inline tests/e2e/volumecreate_shared_test.go]: Done.

**Comment by lipingxue** [comment]
@luomiao I have addressed your comment and please review it.

**Comment by luomiao** [comment]
LGTM

**Pull request merged**